src/ebml.rs

```diff
@@ -134,214 +145,114 @@ pub fn vid(input: &[u8]) -> IResult<&[u8], u64, Error> {
-        let (i, size) = vint(i)?;
-        parse_uint_data(i, size)
+        flat_map(pair(verify(vid, |val| *val == id), vint), |(_, size)| {
+            second(size)
+        })(i)
     }
 }
 
-#[macro_export]
-macro_rules! ebml_int (
-  ($i: expr, $id:expr) => ({
-    $crate::ebml::ebml_int($id)($i)
-  })
-);
-
-pub fn ebml_int(id: u64) -> impl Fn(&[u8]) -> IResult<&[u8], i64, Error> {
-    move |i| {
-        let (i, _) = verify(vid, |val: &u64| *val == id)(i)?;
-        let (i, size) = vint(i)?;
-        parse_int_data(i, size)
-    }
+pub fn ebml_uint<'a>(id: u64) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], u64, Error> {
+    ebml_support(id, parse_uint_data)
 }
 
-#[macro_export]
-macro_rules! ebml_float (
-  ($i: expr, $id:expr) => ({
-    $crate::ebml::ebml_float($id)($i)
-  })
-);
-
-pub fn ebml_float(id: u64) -> impl Fn(&[u8]) -> IResult<&[u8], f64, Error> {
-    move |i| {
-        let (i, _) = verify(vid, |val: &u64| *val == id)(i)?;
-        let (i, size) = vint(i)?;
-        parse_float_data(i, size)
-    }
+pub fn ebml_int<'a>(id: u64) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], i64, Error> {
+    ebml_support(id, parse_int_data)
 }
 
-#[macro_export]
-macro_rules! ebml_str (
-  ($i: expr, $id:expr) => ({
-    $crate::ebml::ebml_str($id)($i)
-  })
-);
-
-pub fn ebml_str(id: u64) -> impl Fn(&[u8]) -> IResult<&[u8], String, Error> {
-    move |i| {
-        let (i, _) = verify(vid, |val: &u64| *val == id)(i)?;
-        let (i, size) = vint(i)?;
-        parse_str_data(i, size)
-    }
+pub fn ebml_float<'a>(id: u64) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], f64, Error> {
+    ebml_support(id, parse_float_data)
 }
 
-#[macro_export]
-macro_rules! ebml_binary (
-  ($i: expr, $id:expr) => ({
-    $crate::ebml::ebml_binary($id)($i)
-  })
-);
-
-pub fn ebml_binary(id: u64) -> impl Fn(&[u8]) -> IResult<&[u8], Vec<u8>, Error> {
-    move |i| {
-        let (i, _) = verify(vid, |val: &u64| *val == id)(i)?;
-        let (i, size) = vint(i)?;
-        parse_binary_data(i, size)
-    }
+pub fn ebml_str<'a>(id: u64) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], String, Error> {
+    ebml_support(id, parse_str_data)
 }
 
-#[macro_export]
-macro_rules! ebml_binary_ref (
-  ($i: expr, $id:expr) => ({
-    $crate::ebml::ebml_binary_ref($id)($i)
-  })
-);
+pub fn ebml_binary<'a>(id: u64) -> impl Fn(&'a [u8]) -> IResult<&'a [u8], Vec<u8>, Error> {
+    ebml_support(id, parse_binary_data)
+}
 
 pub fn ebml_binary_ref(id: u64) -> impl Fn(&[u8]) -> IResult<&[u8], &[u8], Error> {
     move |i| {
-        let (i, _) = verify(vid, |val: &u64| *val == id)(i)?;
-        let (i, size) = vint(i)?;
-        take(usize::try_from(size).map_err(|_| {
-            error!("Not possible to convert size into usize");
-            nom::Err::Error(custom_error(i, 0))
-        })?)(i)
+        pair(verify(vid, |val| *val == id), vint)(i)
+            .and_then(|(i, (_, size))| take(usize_error(i, size)?)(i))
     }
 }
 
-#[macro_export]
+/*#[macro_export]
```

Comment:
Why leave this here if you re-implement it as function below? Also for others further below